internal/service/alarms/internal/db/repo/alarms_repository.go

```diff
@@ -51,3 +59,193 @@ func (ar *AlarmsRepository) GetAlarmEventRecordWithUuid(ctx context.Context, uui
+
+	// Add the values for each record
+	values := make([]bob.Mod[*dialect.InsertQuery], 0, len(records))
+	for _, record := range records {
+		// Build the values for the row
+		args := []any{
+			record.AlarmRaisedTime, record.AlarmChangedTime, record.AlarmClearedTime,
+			record.AlarmAcknowledgedTime, record.AlarmAcknowledged, record.PerceivedSeverity,
+			record.Extensions, record.ResourceID, record.ResourceTypeID, record.NotificationEventType,
+			record.AlarmStatus, record.Fingerprint, record.AlarmDefinitionID, record.ProbableCauseID,
+		}
+
+		if hasAlarmEventRecordID {
+			if record.AlarmEventRecordID != nil {
+				args = append([]any{record.AlarmEventRecordID}, args...)
+			} else {
+				args = append([]any{nil}, args...)
+			}
+		}
+
+		values = append(values, im.Values(psql.Arg(args...)))
+	}
+
+	// Apply the values to the query
+	query.Apply(values...)
+
+	// constrains only applicable for non-archive table
+	if tableName == models.AlarmEventRecordView.Name(context.Background()).String() {
+		query.Apply(im.OnConflictOnConstraint("unique_fingerprint_alarm_raised_time").DoUpdate(
+			im.SetExcluded("alarm_changed_time"),
+			im.SetExcluded("alarm_cleared_time"),
+			im.SetExcluded("alarm_status"),
+			im.SetExcluded("resource_id"),
+			im.SetExcluded("alarm_definition_id"),
+			im.SetExcluded("probable_cause_id"),
+		))
+	}
+
+	// Compile
+	sql, args, err := query.Build()
+	if err != nil {
+		return sql, args, err
+	}
+
+	return sql, args, nil
+}
+
+// GetAlarmDefinitions needed to build out aer
+func (ar *AlarmsRepository) GetAlarmDefinitions(ctx context.Context, am *api.AlertmanagerNotification) ([]models.AlarmDefinition, error) {
+	query := psql.Select(
+		sm.Columns("alarm_name", "alarm_definition_id", "probable_cause_id", "resource_type_id"),
+		sm.From("alarm_definition"),
+		sm.Where(
+			psql.Group(psql.Quote("alarm_name"), psql.Quote("resource_type_id")).
+				In(getGetAlertNameAndResourceTypeID(am)...), // Dynamically pass the pairs
+		),
+	)
+
+	sql, params, err := query.Build()
+	if err != nil {
+		return []models.AlarmDefinition{}, fmt.Errorf("failed to build alarm definitions query when processing AM notification: %w", err)
+	}
+
+	// Run query
+	rows, _ := ar.Db.Query(ctx, sql, params...)
+	records, err := pgx.CollectRows(rows, pgx.RowToStructByNameLax[models.AlarmDefinition])
+	if err != nil {
+		return []models.AlarmDefinition{}, fmt.Errorf("failed to collect alarm definitions: %w", err)
+	}
+
+	return records, nil
+}
+
+func getGetAlertNameAndResourceTypeID(am *api.AlertmanagerNotification) []bob.Expression {
+	var b []bob.Expression
+	for _, alert := range am.Alerts {
+		name := alertmanager.GetAlertName(*alert.Labels)
+		resourceTypeId := alertmanager.GetResourceTypeID(*alert.Labels)
+		b = append(b, psql.ArgGroup(name, resourceTypeId))
+	}
+
+	return b
+}
+
+// ArchiveResolvedAlarmEventRecord Delete resolved and move to archive
+func (ar *AlarmsRepository) ArchiveResolvedAlarmEventRecord(ctx context.Context) error {
```

Comment:
This should be run in a transaction otherwise we run the risk of having an inconsistent view of the data... if the process dies or is restarted at exactly between lines 205 and 206 then we'll have deleted data but not moved it to the archive.

Reply:
Yeah actually a great idea! I was actually suffering from it during dev (it was deleting but wasn't inserting cause of an unrelated error) and was gonna look into transaction before posting the PR but totally forgot lol

Reply:
updated to use transaction!